testutils/src/main/java/com/microsoft/identity/internal/testutils/nativeauth/api/TemporaryEmailService.kt

```diff
@@ -57,32 +67,47 @@ class TemporaryEmailService {
      * To cater for email send and arrival delays, this is retried several times.
      */
     fun retrieveCodeFromInbox(emailAddress: String): String {
-        var retrievedMailbox = api.retrieveMailbox(emailAddress)
-
-        var mailBoxEmpty = retrievedMailbox.isEmpty()
+        var validCodeRetrieved = false
         var count = 0
-        while (mailBoxEmpty && count < 3) {
-            // Wait 3 seconds
-            Thread.sleep(3000)
-
-            // Attempt to retrieve mailbox again.
-            retrievedMailbox = api.retrieveMailbox(emailAddress)
-            mailBoxEmpty = retrievedMailbox.isEmpty()
-
-            // Max 3 retries
-            count++
+        var latestEmailId: String?
+        var otpValue = ""
+        var apiException: Exception? = null
+
+        while (count < numberOfRetries && !validCodeRetrieved) {
+            try {
+                //wait before calling the email endpoint
+                Thread.sleep(retryDelayMs)
+
+                //API returns dates in the UTC timezone, so system time should also be converted
+                val currentTime = ZonedDateTime.ofInstant(Instant.now(), ZoneOffset.UTC)
+
+                val inboxEmails = api.retrieveMailbox(emailAddress)
+                val newEmailId = inboxEmails
+                    .filter { Duration.between(LocalDateTime.parse(it.date, dateFormatter), currentTime).seconds < newEmailCutoff }
+                    .map { it.id }
+                    .firstOrNull()
+
+                if (newEmailId != null) {
+                    latestEmailId = newEmailId
+
+                    val emailContent = api.retrieveEmail(emailAddress, latestEmailId)
+                    otpValue = retrieveOtpFromEmailBody(emailContent.textBody)
+                    validCodeRetrieved = true
```

Comment:
we don't know if the code is valid, right? Not until the API has accepted it.

Reply:
Yes, you are right. The ideal definition of valid code is the code accepted by the API. The current definition of valid code here is if the service successfully filters out/finds the code.

The temp email service is a whole thing that we use to get otp.
```
val otp = tempEmailApi.retrieveCodeFromInbox(user)
val submitCodeResult = (signInResult as SignInResult.CodeRequired).nextState.submitCode(otp)
```
If we can't get past step one, there's no way to implement and validate step two.

Thus I personally find it hard to realize the former definition.